Feature/Packet capture using tshark processing of tcpdump capture

Galvien opened the pull request:
As a third attempt, this change allows easy, **fast** and simple analysis of CMDU messages passing through the container's bridge interface.

The general idea is that each running test is monitored by tcpdump, then saves the packet data into a pcap file, which is later quickly scanned by tshark.

In addition, a helper method extends the dictionary with direct, casted entries of more important values from the tree - for easier access and handling.

Commit:
tests/test_flows: enable tcpdump by default

As preparation for the new message validation methodology using tshark
analysis of tcpdump captures, tcpdump has to be enabled by default.
Therefore, remove the tcpdump option and initialize sniffing on every
run.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tests/sniffer: call tcpdump command with -U

The -U flag makes sure that the pcap file is flushed after every packet
received. Without this flag, the file is buffered, so we will not be
able to read the pcap file until either the tcpdump process is
terminated or the buffer is full. Since later on we will read from the
pcap file in the middle of the tests, we need it to be flushed.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tests/sniffer: add tshark capture and parse method

Add a method parsing the pcap file into json using tshark, deserializing
it with python.

We need to know the file to which we are capturing for this. Therefore,
store it in the Sniffer object, and reset it when the capture is
stopped. Note that the latter makes it impossible to use the
get_packet_capture() method after the test - for now, that's how we
intend to use it, but this may change in the future.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
tests/sniffer: convert sniffer results to Packet

The JSON output from tshark is a bit unwieldy. Introduce a Packet class
that hoists some of the useful fields to the top level, and exposes them
as attributes so no dictionary lookup is needed for frequently used
attributes. We're mainly concerned with ieee1905, so these are exposed
in detail.

Similarly, recursively reorganise the TLVs into Tlv structures.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Galvien commented:
@arnout Gitlab CI will always fail since this change completely removes the `--tcpdump` option.
I think we'll need to change that.

tomereli commented:
> @arnout Gitlab CI will always fail since this change completely removes the `--tcpdump` option.
> I think we'll need to change that.

So change what you need in .gitlab-ci.yml

tomereli commented on tests/opts.py:
```diff
@@ -9,7 +9,6 @@
 class opts:
     '''Static class that encodes the global options.'''
     verbose = False
```
If you remove this section, you need to also update the README to mention that tcpdump is now a requirement for running tests even in docker. You should also explain how to fix the capabilities so that non sudo users can use tcpdump.

tomereli commented on tests/sniffer.py:
```diff
@@ -13,30 +13,33 @@
+    def start(self, test_name):
+        '''Start tcpdump.'''
+        outputfile_basename = "test_" + test_name
+        debug("Starting tcpdump, output file {}.pcap".format(outputfile_basename))
```
I don't think you should make the output file hard coded - why is this needed? You can just add another argument with default value (None) for the output file name - so that if it is None, you will use the test name.

arnout replied:
When the outputfiles dict is removed, this change is not necessary.

tomereli commented on tests/sniffer.py:
```diff
@@ -26,7 +26,7 @@ def start(self, test_name):
         outputfile = os.path.join(opts.tcpdump_dir, outputfile_basename) + ".pcap"
         self.outputfiles[test_name] = outputfile
-        command = ["tcpdump", "-i", self.interface, '-w', outputfile]
+        command = ["tcpdump", "-i", self.interface,'-U', '-w', outputfile]
```
I don't understand whNot sure I understand that the `-U` flag is used for here:

From tcpdump manual:

>-U
--packet-buffered
If the -w option is not specified, or if it is specified but the --print flag is also specified, make the printed packet output ``packet-buffered''; i.e., as the description of the contents of each packet is printed, it will be written to the standard output, rather than, when not writing to a terminal, being written only when the output buffer fills.
If the -w option is specified, make the saved raw packet output ``packet-buffered''; i.e., as each packet is saved, it will be written to the output file, rather than being written only when the output buffer fills.
The -U flag will not be supported if tcpdump was built with an older version of libpcap that lacks the pcap_dump_flush(3PCAP) function.

Your commit message also doesn't explain what issue this flag is solving. Please elaborate.

arnout replied:
Without -U, the output file will be buffered. That means we'd have to terminate tcpdump to see the last few packets in the output file. with the -U option, it is packet-buffered, i.e. whenever a packet is captured, the file is flushed. That way, when we read the file, we're sure that it contains everything that has been captured.

tomereli commented on tests/sniffer.py:
```diff
@@ -41,6 +42,23 @@ def start(self, test_name):
+            tshark_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        try:
+            capture = json.loads(self.tshark_proc.stdout)
+            def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
```
I don't think this is appropriate - instead, we should capture only ieee1905 packets from the start when we use tcpdump:
```
tcpdump -i<iface> ether proto 0x893a
```

tomereli commented on tests/sniffer.py:
```diff
@@ -41,6 +42,23 @@ def start(self, test_name):
+            self.captures[test_name] = result
+            return self.captures[test_name]
+        except BaseException as ex:
+            return None
```
Can you explain what kind of exception we're ignoring here?

tomereli commented on tests/sniffer.py:
```diff
@@ -59,6 +59,20 @@ def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
         except BaseException as ex:
             return None
 
+    def _simplify_packets(self, packets):
```
This should be reordered since you use this method in the previous commit.

tomereli commented on tests/test_flows.py:
```diff
@@ -365,10 +365,12 @@ def test_ap_capability_query(self):
         debug("Confirming ap capability query has been received on agent")
         self.check_log(env.agents[0], "AP_CAPABILITY_QUERY_MESSAGE")
-
+        capability_query = env.wired_sniffer.get_packets_for_cmdu_type(self.running,0x8001)
```
The example should replace the check_log, not just add a call without testing it. Also I think you should add an API to search for specific packets like it is done in the UCC.

For example, this is how the UCC is checking for a specific topology response:

```
2020-04-14 04:08:41.506 - INFO - 
        ~~~~~ 4.1.1 Step 9 sniffer checks - Topology Response message ~~~~~

2020-04-14 04:08:41.512 - INFO - SNIFFER (192.168.250.5:9999) ---> sniffer_control_filter_capture,InFile,_MAP-4.1.1-STEP9,OutFile,MAP-4.1.1-STEP9,msgId,0x43c6,filterfile,mapfile,msgType,0x0003,nframes,1
2020-04-14 04:08:42.023 - INFO - SNIFFER (192.168.250.5:9999) <-- status,COMPLETE,FilterStatus,SUCCESS,FrameCount,1

```

I think the equivalent in python should be something like:

```
topology_response = env.wired_sniffer.get_packets(self.running, cmdu_type=0x0003, mid=0x43c6)
```
The API should also allow for more arguments - source and destination MACs so we can filter on ALid, and probably more that I can't think of now. Every parameter should be optional except for self.running.

tomereli commented on tests/sniffer.py:
```diff
@@ -73,6 +74,23 @@ def _simplify(x):
             return x
         return list(map(_simplify, packets))
 
+    def get_packets_for_cmdu_type(self, test_name, msg_type: int):
```
I think we're missing an important argument here - message identifier (mid) - this one should be used to verify that we are looking at the correct cmdu. This should be optional parameter, so if its none, it should return all the CMDUs which match the message type, otherwise it would return a single CMDU or None if there is no message with this MID.

In addition, we must also add source and destination macs parameters, since this is the only way we can uniquely identify a message - source + destination ALid + mid.

arnout replied:
Yeah, there are multiple ways to go about it. Basically we want to check "was there a packet that matches these and these parameters".

I think actually the best way to do it would be to have a check_packet function in TestFlows itself that takes the filter as an argument. The filter can then check all the data it needs to.

tomereli reviewed (changes requested):
Very nice PR, good commits split. I have some comments so re-review is required.

arnout commented on tests/sniffer.py:
```diff
@@ -7,36 +7,89 @@
     def __init__(self, interface: str):
         self.interface = interface
         self.tcpdump_proc = None
+        self.outputfiles = {}
```
There is no need for this dictionary. At any point, there is only one running capture, thus only one running outputfile. All the get_packet_capture() calls will be for that last outputfile.

Note that multiple captures in parallel is not possible because they're all capturing the same packets.

arnout commented on tests/sniffer.py:
```diff
@@ -18,6 +18,7 @@ def __init__(self, interface: str):
         self.interface = interface
         self.tcpdump_proc = None
         self.outputfiles = {}
+        self.captures = {}
```
Again, not needed: just use the latest one.

arnout commented on tests/sniffer.py:
```diff
@@ -59,6 +59,20 @@ def ieee_filter(x): return 'ieee1905' in x['_source']['layers']
 
+    def _simplify_packets(self, packets):
+        def _simplify(x):
+            x['src_mac'] = x['_source']['layers']['eth']['eth.src']
```
(nitpick) Instead of "simplifying", it would be nicer to convert to a brand new class Ieee1905Message. Or a namedtuple instead o a full class.

arnout committed:
tests: sniffer: remove use_tcpdump option

Since the two places where the sniffer is used (test_flows environment
and boardfarm) are both enabling the use_tcpdump option unconditionally,
remove the option.

Update the callers accordingly.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: sniffer: add checkpoint function

The checkpoint function saves the frame number of the last capture
packet. Subsequent calls to get_packet_capture start from that frame.
This is done directly with a tshark display filter.

This approach is not the most efficient one, since tshark is called
multiple times. But since this is just testing and not very intensive,
it's OK like this.

Add a checkpoint function in the environment module as well. For now, it
just checkpoints the sniffer, but in the future it will do the same for
the log files.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows: add function check_cmdu

This function checks if an IEEE1905 CMDU with the given features was
captured and fails if it was not.

Note that get_packet_capture is called every time, which spawns a tshark
process, which is not terribly efficient. However, since this is only
for testng and anyway not done 1000s of times, it's acceptable.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: test_channel_selection: use check_cmdu_type

For all checks that were merely verifying that a packet arrived, replace
them with check_cmdu_type calls. While we're at it, also verify all the
MID values, and extend the checks a little to cover more ground.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: sniffer: capture only LLDP and IEEE1905.1 packets

Since there may be a lot more traffic than what we're interested in,
add a filter to tcpdump to include only what we are going to check: LLDP
and IEEE1905.1 packets.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tools/docker/README.md: tcpdump and tshark are needed for running tests

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout reviewed (approved):
I took over this PR and fixed all the issues that were reported in this one and in #1020.

Now it is perfect :-)

arnout committed:
docker: tests-runner: add tshark

Since the tests now require tshark, add tshark to the docker image used
for tests.

While we're at it, also fully specify the registry from which we pull
the dind image. This avoids problems on systems that have a different
registry configured as their default. In a similar vein, specify the
image by digest instead of by tag. This way, the image is 100%
reproducible.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

rmelotte commented on tests/sniffer.py:
```diff
@@ -23,7 +23,7 @@ def start(self, outputfile_basename):
         os.makedirs(os.path.join(self.tcpdump_log_dir, 'logs'), exist_ok=True)
         outputfile = os.path.join(self.tcpdump_log_dir, outputfile_basename) + ".pcap"
-        command = ["tcpdump", "-i", self.interface, "-w", outputfile]
+        command = ["tcpdump", "-i", self.interface, '-U', '-w', outputfile]
```
Maybe it would be worth adding the `-s` option with the maximum size of a wireless frame (2346).
The default size depends on the tcpdump version, but on my machine for example it's set to 262144, while we know the maximum size we'll ever capture can't exceed 2346 (and probably even 1516 in practice).

AFAIU tcpdump sets it so high for backward compatibility, but they recommend to set it to something smaller (from man tcpdump):

> You should limit snaplen to the smallest number that will  capture the protocol information you're interested in.

(it's a nitpick comment of course, we shouldn't have that much traffic for now, so we can still do that once we start sniffing wireless frames)

arnout replied:
I believe snaplen only has an impact on memory usage, not so much on performance. So I'm not really worried about it. That said, it's a good suggestion.

BTW I intend to also drop tcpdump and replace it with tshark at some point - it's kind of stupid to require both to be installed.

rmelotte reviewed (approved):
I didn't review the details, overall LGTM.

I only have a nitpick comment, but it will probably only become relevant once we start sniffing wireless interfaces, so I'll add the "ready for merge" tag anyway.
